### ISSUE TITLE
Regular Expressions I Have Known and Loathed

### ISSUE THREAD

**Issue by bernie-simon** [spacetelescope/braindump#22]
Whenever I mention regular expressions, I get the impression that some people are uncomfortable with them. Perhaps this is too basic, but I can give a talk on regular expressions, using as an example parsing words out of a line of text, building up from a simple regular expression to something that is hideous to contemplate.

**Comment by perrygreenfield** [comment]
Good idea, though I’m aways reminded the quote: "Some people, when confronted with a problem, think "I know, I'll use regular expressions." Now they have two problems.”

But seriously, they do have a really important role as long as one doesn’t try to do too much with them.

**Comment by bernie-simon** [comment]
Talk will use an example in Python. Makes little difference, as the regular expressions to be discussed are the same in Perl and Python.

**Comment by hcferguson** [comment]
I always found regular expressions to be really easy (e.g. in awk) -- until I tried to use them in Python. Now I cringe when I find myself with a problem requiring re's.

**Comment by jdavies-st** [comment]
https://regex101.com is my friend when trying to write, change or debug regular expressions.

**Comment by pllim** [comment]
Here is how I write regex: I go to James or Jonathan's office and have them write it for me.

**Comment by drdavella** [comment]
Maybe someone should give a talk on [using regular expressions to parse html](https://stackoverflow.com/questions/1732348/regex-match-open-tags-except-xhtml-self-contained-tags/1732454#1732454)...

**Comment by mdlpstsci** [comment]
I have only had to use them in a few instances so I am rusty, but I LOVE regular expressions.  This might be because I once had to learn Standard ML for a class.  

Please sir, may I have some more?  I vote fof this talk

**Comment by perrygreenfield** [comment]
But it probably could be used as part of a lexer in a real parser. By itself, madness.

**Comment by sosey** [comment]
(go cats!) maybe we should start some local usefull re example list somewhere for people to reference, besides just the talk information?

I'm not saying this is a good example, just an example :)

```
token = '\s+|(?<!\d)[,](?!\d)'
letters = re.compile("(^[a-zA-Z])")  # starts with a letter
numbers = re.compile("(^(?:[+\-])?(?:\d*)(?:\.)?(?:\d*)?(?:[eE][+\-]?\d*$)?)")
empty = re.compile("(^\s*$)")  # is a blank line

use:
numbers.fullmatch(line)
if not empty.match(line):
            if letters.match(line):
                pair = re.split(token, line.strip(), maxsplit=3)
                if len(pair) == 2:  # key and value exist
                    key = pair[0]  # first item is the key
                    val = pair[1]  # second item is the value
                    if letters.match(val):
                        value = val
                    if numbers.fullmatch(val):
                        value = eval(val)
                if len(pair) == 3:  # key min max exist
                    key = pair[0]
                    val1, val2 = pair[1:]
                    if numbers.fullmatch(val1) and numbers.fullmatch(val2):
                        value = (eval(val1), eval(val2))
                    else:
                        raise ValueError("Min/max values expected for {0}"
                                         .format(key))
```

**Comment by mdlpstsci** [comment]
I think Megan's idea is a good one.   Yay to code sharing.

(I only hope UofA does not mind my using their logo!)


Michele D. De La Pena
Space Telescope Science Institute
mdelapena@stsci.edu
667.218.6501
